lxd/certificates.go

```diff
@@ -601,59 +600,58 @@ func certificatesPost(d *Daemon, r *http.Request) response.Response {
-			dbCert := dbCluster.Certificate{
-				Fingerprint: shared.CertFingerprint(cert),
-				Type:        dbReqType,
-				Name:        name,
-				Certificate: string(pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE", Bytes: cert.Raw})),
-				Restricted:  req.Restricted,
-			}
-
-			_, err := dbCluster.CreateCertificateWithProjects(ctx, tx.Tx(), dbCert, req.Projects)
-			return err
-		})
-		if err != nil {
-			return response.SmartError(err)
+	err = s.DB.Cluster.Transaction(context.TODO(), func(ctx context.Context, tx *db.ClusterTx) error {
+		// Check if we already have the certificate.
+		existingCert, _ := dbCluster.GetCertificateByFingerprintPrefix(ctx, tx.Tx(), fingerprint)
+		if existingCert != nil {
+			return api.StatusErrorf(http.StatusConflict, "Certificate already in trust store")
 		}
 
-		// Notify other nodes about the new certificate.
-		notifier, err := cluster.NewNotifier(s, s.Endpoints.NetworkCert(), s.ServerCert(), cluster.NotifyAlive)
-		if err != nil {
-			return response.SmartError(err)
+		// Store the certificate in the cluster database.
+		dbCert := dbCluster.Certificate{
+			Fingerprint: shared.CertFingerprint(cert),
+			Type:        dbReqType,
+			Name:        name,
+			Certificate: string(pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE", Bytes: cert.Raw})),
+			Restricted:  req.Restricted,
 		}
 
-		req := api.CertificatesPost{
-			CertificatePut: api.CertificatePut{
-				Certificate: base64.StdEncoding.EncodeToString(cert.Raw),
-				Name:        name,
-				Type:        api.CertificateTypeClient,
-			},
-		}
+		_, err := dbCluster.CreateCertificateWithProjects(ctx, tx.Tx(), dbCert, req.Projects)
+		return err
+	})
+	if err != nil {
+		return response.SmartError(err)
+	}
 
-		err = notifier(func(client lxd.InstanceServer) error {
-			return client.CreateCertificate(req)
-		})
-		if err != nil {
-			return response.SmartError(err)
-		}
+	// Send a notification to other cluster members to refresh their identity cache.
+	notifier, err := cluster.NewNotifier(s, s.Endpoints.NetworkCert(), s.ServerCert(), cluster.NotifyAlive)
+	if err != nil {
+		return response.SmartError(err)
+	}
 
-		// Add the certificate resource to the authorizer.
-		err = s.Authorizer.AddCertificate(r.Context(), fingerprint)
-		if err != nil {
-			logger.Error("Failed to add certificate to authorizer", logger.Ctx{"fingerprint": fingerprint, "error": err})
-		}
+	req = api.CertificatesPost{
+		CertificatePut: api.CertificatePut{
+			Certificate: base64.StdEncoding.EncodeToString(cert.Raw),
+			Name:        name,
+			Type:        api.CertificateTypeClient,
+		},
 	}
 
-	// Reload the cache.
+	err = notifier(func(client lxd.InstanceServer) error {
+		_, _, err := client.RawQuery(http.MethodPost, "/internal/identity-cache-refresh", nil, "")
+		return err
+	})
+	if err != nil {
+		return response.SmartError(err)
+	}
+
+	// Add the certificate resource to the authorizer.
+	err = s.Authorizer.AddCertificate(r.Context(), fingerprint)
```

Comment:
What does this do?

Reply:
These were not removed when we reverted the original OpenFGA authorization driver because they were part of a separate pull request that performed a lot of refactoring. The no longer do anything and I will remove them when the embedded driver is in.

Reply:
https://github.com/canonical/lxd/issues/12975

Reply:
Thanks, glad I mentioned it :)